recovery/elliptics_recovery/dc_server_send.py

```diff
@@ -78,6 +79,81 @@ def __init__(self, bucket_file, group_id):
         self.bucket_file = bucket_file
         self.group_id = group_id
 
+    def sort_by_physical_order(self):
+        try:
+            # phase 1: sort chunks
+            max_chunks_size = 500 * 1024 * 1024 # 500 Mb
+
+            chunk_files = []
+            chunk = []
+            chunk_size = 0
+
+            self.bucket_file.seek(0)
+            for key_data in load_key_data_from_file(self.bucket_file):
+                chunk_size += sys.getsizeof(key_data)
+                chunk.append(key_data)
+                if chunk_size > max_chunks_size:
+                    chunk_files.append(self._create_chunk_file(len(chunk_files)))
+                    self._sort_chunk(chunk, chunk_files[-1])
+                    chunk = []
+                    chunk_size = 0
+            if chunk:
+                chunk_files.append(self._create_chunk_file(len(chunk_files)))
+                self._sort_chunk(chunk, chunk_files[-1])
+
+            # phase 2: merge chunks
+            class MergeDataWrapper(object):
+                def __init__(self, gen, outer):
+                    self.gen = gen
+                    self.outer = outer
+                    self.next()
+
+                def __cmp__(self, other):
+                    return cmp(self.key_data[0], other.key_data[0])
+
+                def next(self):
+                    try:
+                        self.key_data = self.outer._extend_key_data(self.gen.next())
+                    except StopIteration:
+                        return False
+                    return True
+
+            import heapq
+            heap = []
+            for chunk in chunk_files:
+                chunk.seek(0)
+                heapq.heappush(heap, MergeDataWrapper(load_key_data_from_file(chunk), self))
+
+            self.bucket_file.seek(0)
+            self.bucket_file.truncate()
+            while len(heap):
+                wrapper = heapq.heappop(heap)
+                dump_key_data(wrapper.key_data[1], self.bucket_file)
+                if wrapper.next():
+                    heapq.heappush(heap, wrapper)
+        finally:
+            for f in chunk_files:
+                os.unlink(f.name)
+
+    def _extend_key_data(self, key_data):
```

Comment:
You don't need `key_data` extending, since it is used only for comparison purposes and can be done via comparison of first `key_info` parts from `key_data`. The last is correct because [`key_infos` in `key_data` are sorted in the required order] (https://github.com/interiorem/elliptics/blob/master/recovery/elliptics_recovery/types/dc.py#L253).